### ISSUE TITLE
Can I use it in a very large dataset, eg . million

### ISSUE THREAD

**Issue by presburger** [angus924/hydra#4]

**Comment by angus924** [comment]
Hi @Presburger.

Yes, you can, but you will need to consider the best way to do this efficiently.

The best approach is probably to train a logistic/softamax regression classifier using sgd/adam (as in [`code/softmax.py`](https://github.com/angus924/hydra/blob/main/code/softmax.py)).

I would suggest performing the (Hydra) transform in "chunks" and saving the transformed features to disk.  With 1 million training examples, this might use 30 or 40 gb of disk space (this will depend on the length of the time series and how many features are produced by the transform).  (Of course, if you have enough ram, you can just store the transformed features in memory.)  This way you only need to transform the input time series once (not over and over again, e.g., every epoch of training).  You can either do this as a separate step, or integrate it into minibatch sgd/adam training.  One way to do this is using numpy memmap, so something like this:

```python
cache = np.memmap(path, mode = "w+", dtype = "float32", shape = (num_examples, num_features))

chunks = torch.arange(num_examples).split(chunk_size)

for chunk in chunks:
  cache[chunk] = transform(training_data[chunk])
```

You will need to think about whether your training data is "shuffled" (i.e., whether you want to also shuffle the training data or not before you perform the transform).  When you are reading the transformed features from disk (i.e., in your sgd training loop), you need to be careful to read contiguous chunks from disk, otherwise it can get pretty slow.

I'm working on some updated code to train on large datasets (updating `softmax.py`), hopefully that will be ready soon.

**Comment by presburger** [comment]
@angus924 thanks for the reply.
I have enough ram, But I read the code, i find
Select 10000 sample determine bias from very large data set, it course great randomness.

**Comment by angus924** [comment]
> Select 10000 sample determine bias from very large data set

Sorry, I'm not quite sure what you mean.  Do you mean that the size of the validation set is too small?

Or do you mean something else?

**Comment by presburger** [comment]
> > Select 10000 sample determine bias from very large data set
> 
> Sorry, I'm not quite sure what you mean. Do you mean that the size of the validation set is too small?
> 
> Or do you mean something else?

 hi,
The convlution bias ​​of each selection are quite different.
I read the code, minirocket will select feature size(eg. 10000) sample from the datasets.
so cause the great randomless of the accurency of the model .

**Comment by angus924** [comment]
Sorry, now I understand, I thought you were talking about Hydra.

With MiniRocket, yes, by default we randomly draw examples from the training set to find bias values.

With a big dataset, you need to use a sufficiently-large and representative subset of the training set when calling `fit(...)`.  (For example, if you just use the first 100, or 1,000, or 10,000 time series in the dataset, and these haven't been shuffled, it's unlikely to work properly.)

If you call `fit(...)` with different subsets of, let's say, 10,000 training examples, and each of those subsets is a good, well-shuffled, representative subset of the training data, and you are still getting very different results for / high variance between each of those subsets, to me that suggests that there's quite a lot of variance in the input time series.

When you call `fit(...)`, you are basically just trying to collect values which are representative of the distribution of the values in the convolution output (over the training set).  If a sample of 10,000 is insufficient to do this, it indicates that the values in the convolution output are changing a lot between different samples of 10,0000 from the training set.

The first thing to make sure would be that the subset of the training set that you are using when calling `fit(...)` is sufficiently large and well-shuffled, etc.  (I think in [`softmax.py`](https://github.com/angus924/minirocket/blob/main/code/softmax.py) by default we sample from the first 4,096 examples, that's probably not enough for a training set of size 1 million, you might want to make that bigger.)

The simplest approach to sampling bias values more broadly would be to choose a subset of, say, 10,000, or 20,000, or 50,000 time series, transform them using [`minirocket_dv.py`](https://github.com/angus924/minirocket/blob/main/code/minirocket_dv.py) (it transforms the data and samples the bias values at the same time, so  you don't have to repeat the transform twice), and then transform the rest using the standard `transform(...)` function from [`minirocket.py`](https://github.com/angus924/minirocket/blob/main/code/minirocket.py).

If that's too slow, but you still want to sample bias value from a lot more input time series, then you could consider modifying the code slightly to sample multiple input time series when finding bias values.  You could modify either [`minirocket.py`](https://github.com/angus924/minirocket/blob/main/code/minirocket.py) or maybe it would be easier to work from [`minirocket_dv.py`](https://github.com/angus924/minirocket/blob/main/code/minirocket_dv.py).

Something like this ([`minirocket_dv.py`](https://github.com/angus924/minirocket/blob/main/code/minirocket_dv.py), from line 64):

```python
[...]

num_examples_per_bias_value = 16 # this would be an input argument

[...]

C = np.zeros((num_examples_per_bias_value, input_length), dtype = np.float32)

sample_indices = np.random.randint(0, num_examples, num_examples_per_bias)

for example_index in sample_indices:

    _X = X[example_index]

    A = -_X          # A = alpha * X = -X
    G = _X + _X + _X # G = gamma * X = 3X

[...]
```

...something like that.

So, here, you are sampling 16 time series, transforming them, and then taking the bias values from the output for all 16 time series.  (I haven't tested the above code, but it shows the general idea.)  This would allow you to use a large sample of input time series and get good coverage over the subset without having to transform everything.

Does that help?

**Comment by presburger** [comment]
@angus924 
Great, it is a great idea. thanks very much.
